cmd/fsck/main.go

```diff
@@ -0,0 +1,266 @@
+	N           = flag.Uint("N", 1, "The number of workers to use when fetching/comparing resources")
+	origin      = flag.String("origin", "", "Origin of the log to check")
+	pubKey      = flag.String("public_key", "", "The log's public key in base64 encoded DER format")
+)
+
+func main() {
+	klog.InitFlags(nil)
+	flag.Parse()
+	ctx := context.Background()
+	logURL, err := url.Parse(*storageURL)
+	if err != nil {
+		klog.Exitf("Invalid --storage_url %q: %v", *storageURL, err)
+	}
+	hc := &http.Client{
+		Transport: &http.Transport{
+			MaxIdleConns:        int(*N),
+			MaxIdleConnsPerHost: int(*N),
+			DisableKeepAlives:   false,
+		},
+		Timeout: 30 * time.Second,
+	}
+	src, err := client.NewHTTPFetcher(logURL, hc)
+	if err != nil {
+		klog.Exitf("Failed to create HTTP fetcher: %v", err)
+	}
+	if *bearerToken != "" {
+		src.SetAuthorizationHeader(fmt.Sprintf("Bearer %s", *bearerToken))
+	}
+	v := verifierFromFlags()
+	if *origin == "" {
+		*origin = v.Name()
+	}
+	lsc := &logStateCollector{}
+	if err := fsck.Check(ctx, *origin, v, src, *N, lsc.merkleLeafHasher()); err != nil {
+		klog.Exitf("fsck failed: %v", err)
+	}
+
+	if err := checkIntermediates(ctx, lsc, src.ReadIssuer); err != nil {
+		klog.Exitf("Failed to verify presence intermediates: %v", err)
+	}
+
+	klog.Info("OK")
+}
+
+func checkIntermediates(ctx context.Context, lsc *logStateCollector, readIssuer func(context.Context, []byte) ([]byte, error)) error {
+	klog.Infof("Checking intermediates CAS")
+	n := 0
+	work := make(chan []byte, 10)
+	eg := errgroup.Group{}
+	for range 10 {
+		eg.Go(func() error {
+			for fp := range work {
+				if _, err := readIssuer(ctx, fp); err != nil {
+					return fmt.Errorf("couldn't fetch issuer for %x: %v", fp, err)
+				}
+			}
+			return nil
+		})
+	}
+
+	lsc.intermediates.Range(func(k any, v any) bool {
+		fp := k.(string)
+		work <- []byte(fp)
+		n++
+		return true
+	})
+	close(work)
+
+	if err := eg.Wait(); err != nil {
+		return fmt.Errorf("failed to fetch one or more issuers: %v", err)
+	}
+
+	klog.Infof("Checked %d intermediate issuers", n)
+	return nil
+}
+
+type logStateCollector struct {
+	intermediates sync.Map
+}
+
+func (l *logStateCollector) addIntermediates(fpRaw cryptobyte.String) {
+	var fp []byte
+	for len(fpRaw) > 0 {
+		fp, fpRaw = fpRaw[:32], fpRaw[32:]
+		_, existed := l.intermediates.LoadOrStore(string(fp), true)
+		if !existed {
+			klog.V(2).Infof("Found intermediate FP %x", fp)
+		}
+	}
+}
+
+// merkleLeafHasher returns a function which knows how to:
+// - calculate RFC6962 Merkle leaf hashes for entries in a Static-CT formatted entry bundle,
+// - keep track of the set of intermediate cert fingerprints seen while parsing entry bundles.
+func (l *logStateCollector) merkleLeafHasher() func(bundle []byte) ([][]byte, error) {
+	return func(bundle []byte) ([][]byte, error) {
+		r := make([][]byte, 0, layout.EntryBundleWidth)
+		b := cryptobyte.String(bundle)
+		for i := 0; i < layout.EntryBundleWidth && !b.Empty(); i++ {
```

Comment:
I think you should be able to use something like https://github.com/transparency-dev/tesseract/blob/024f1e989970d41505590d1c65c3b3e7ed9cba34/internal/types/staticct/staticct.go#L47 or https://github.com/transparency-dev/tesseract/blob/024f1e989970d41505590d1c65c3b3e7ed9cba34/internal/types/staticct/staticct.go#L191 It might not work out of the box, but that's working as intended: I'd like to understand what we need these parsing functions for exactly before having one to rule them all. If you can't find an easy way to use any of these methods, put the parsing method you need over there, and at least they will all be in the same file.

Now, here's an easy counter argument: this is a check tool..... so maybe it *should* reimplement everything?